fuse/readonly/readonly_unix.go

```diff
@@ -144,14 +145,44 @@ func (s *Node) ReadDir(intr fs.Intr) ([]fuse.Dirent, fuse.Error) {
 	return nil, fuse.ENOENT
 }
 
-// ReadAll reads the object data as file data
-func (s *Node) ReadAll(intr fs.Intr) ([]byte, fuse.Error) {
-	log.Debug("Read node.")
-	r, err := uio.NewDagReader(context.TODO(), s.Nd, s.Ipfs.DAG)
+func (s *Node) Read(req *fuse.ReadRequest, resp *fuse.ReadResponse, intr fs.Intr) fuse.Error {
+	// intr will be closed by fuse if the request is cancelled. turn this into a context.
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel() // make sure all operations we started close.
+
+	// we wait on intr and cancel our context if it closes.
+	go func() {
+		select {
+		case <-intr: // closed by fuse
+			cancel() // cancel our context
+		case <-ctx.Done():
+		}
+	}()
+
+	k, err := s.Nd.Key()
+	if err != nil {
+		return err
+	}
+
+	// setup our logging event
+	lm := make(lgbl.DeferredMap)
+	lm["fs"] = "ipfs"
+	lm["key"] = func() interface{} { return k.Pretty() }
+	lm["req_offset"] = req.Offset
+	lm["req_size"] = req.Size
+	defer log.EventBegin(ctx, "fuseRead", lm).Done()
+
+	r, err := uio.NewDagReader(ctx, s.Nd, s.Ipfs.DAG)
```

Comment:
Oh crap, i missed this. We dont need to construct a new reader. The Node object has one in it already.